### DISCUSSION TITLE
How can I return from parsing ahead of time?

### DISCUSSION THREAD

**Discussion by shuizaiku7** [taocpp/PEGTL#342] [Q&A] [answered]
Hi, I'm parsing big files, but somethings I don't need to parse them all.
Maybe I just need to collect a certain number of items from the file.

Is there any mechanism to return from parsing ahead of time?

**Comment by ColinH** [comment]
Sure, your grammar doesn't have to include the `eof` rule in a way that a successful parse finishes with it. Consider this edge case, if your grammar matches a single `a` via `tao::pegtl::one< 'a' >` and does nothing else your parsing run will finish after reading only the first character of the file and, depending on whether the first character exists and is an 'a', will either succeed or fail.

**Reply by shuizaiku7** [reply]
Sorry I don't get it.
My user may pass a argument to me to specify how many items i need to parse.
So i will store this number in State state, once the item parsed reaches the number, i wanna stop the parsing right away. 
I thought there maybe a function such as break() that I can call when i wanna stop ahead of time.

**Reply by ColinH** [reply]
That sounds like you need a custom rule, some `dynamic_rep` that is similar to `tao::pegtl::rep` but takes the number of repetitions from the state instead of a template parameter.

**Reply by ColinH** [reply]
PS: You could throw an exception when you are done, but I wouldn't go down that path since it's neither elegant nor efficient to end a successful parsing run like that. The PEGTL is designed around an exception during parsing being an unrecoverable "global" failure.

**Reply by ColinH** [reply]
Alternatively you could use the `tao::pegtl::until` rule, increment a counter in the action for each item, and check whether the counters match in a separate action attached to the `tao::pegtl::success` rule that you use as termination condition for the `until`. This works if you use a slightly less often used feature, namely that actions can return `bool` instead of `void`; if an action returns `false` then the rule that it is attached to will be considered as not matched, "as if" the rule had returned `false` itself (though in that case the action wouldn't have been called).

**Reply by shuizaiku7** [reply]
Yeah, i agree that using exception to implement a regular logic is not elegant.
I've considered returning bool from action's apply function, but at first i thought it was used for rewinding sth instead of stop match according to my understanding from our documentation. 

Thank you, i will try the seond way.

**Reply by ColinH** [reply]
Whether a rule (or action) returning `false` stops the parsing run or lets it continue somewhere else (in the grammar) depends fully on the position of this rule in, and the structure of, the overall grammar.

Anyhow, glad to have been of help, let me know if you get stuck.

**Comment by ColinH** [comment, marked as answer]
Alternatively you could use the `tao::pegtl::until` rule, increment a counter in the action for each item, and check whether the counters match in a separate action attached to the `tao::pegtl::success` rule that you use as termination condition for the `until`. This works if you use a slightly less often used feature, namely that actions can return `bool` instead of `void`; if an action returns `false` then the rule that it is attached to will be considered as not matched, "as if" the rule had returned `false` itself (though in that case the action wouldn't have been called).
